src/client.ts

```diff
@@ -6,6 +6,32 @@ import {
 } from "vscode-languageclient";
 import { Uri } from "vscode";
 import type { PythonExtension } from "@vscode/python-extension";
+import type { InitializationOptions, ExtensionSettings } from "./common/settings";
+
+// Keys that are handled by the extension and should not be sent to the server
+type ExtensionOnlyKeys = keyof InitializationOptions | keyof ExtensionSettings | "trace";
+
+const EXTENSION_ONLY_KEYS = [
+  // InitializationOptions
+  "logLevel",
+  "logFile",
+  // ExtensionSettings
+  "cwd",
+  "path",
+  "interpreter",
+  "importStrategy",
+  // Client-handled settings
+  "trace",
+] as const satisfies readonly ExtensionOnlyKeys[];
+
+// Compile-time check that all extension-only keys are included
+type AssertAllKeys<
+  T extends readonly ExtensionOnlyKeys[],
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  _Check = ExtensionOnlyKeys extends T[number] ? true : never,
+> = T;
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+type _AssertComplete = AssertAllKeys<typeof EXTENSION_ONLY_KEYS>;
```

Comment:
This took me a while to understand. 


Two alternative options:

a) We change the server's `configuration` request to explicitly list what options it wants instead of requesting the entire `ty` namespace. This has the benefit that it works for all clients and not just VS Code
b) Use a function with typescript's exhaustivness matching


```ts
function isExtensionOnlyKey(key: string): boolean {
  return isExtensionOnlyKeyImpl(key as ExtensionOnlyKeys);
}

function isExtensionOnlyKeyImpl(key: ExtensionOnlyKeys): boolean {
  switch (key) {
    case "logLevel":
    case "logFile":
    case "cwd":
    case "path":
    case "interpreter":
    case "importStrategy":
    case "trace":
      return true;
  }
}
```

I think I'd prefer option a)

Reply:
another alternative with a bit less magic:


```ts
const EXTENSION_ONLY_KEYS = {
  logLevel: true,
  logFile: true,
  cwd: true,
  path: true,
  interpreter: true,
  importStrategy: true,
  trace: true,
} as const satisfies Record<ExtensionOnlyKeys, true>;

function isExtensionOnlyKey(key: string): key is ExtensionOnlyKeys {
  return key in EXTENSION_ONLY_KEYS;
}
```

the IDE is even able to complete missing keys

Reply:
Thank you! Yeah, that's better